crypto/ascon/ascon_aead128.c

```diff
@@ -0,0 +1,436 @@
+#  define ASCON_INLINE __attribute__((always_inline)) inline
+# elif defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
+#  define ASCON_INLINE inline /* C99? */
+# else
+#  define ASCON_INLINE
+# endif
+#else
+# define ASCON_INLINE
+#endif
+
+#define ROR64(x, i) ((x << (64 - i)) | (x >> i))
+
+/**
+ * constant addition layer, NIST SP 800-232 Table 5
+ * 3c 2d 1e 0f f0 e1 d2 c3 b4 a5 96 87 78 69 5a 4b
+ */
+#define ASCONPC(x0, x1, x2, x3, x4, rcon) \
+    do {                                  \
+        x2 ^= rcon;                       \
+    } while (0)
+
+/**
+ * nonlinear layer, lifted from p43 of
+ * https://csrc.nist.gov/CSRC/media/Projects/lightweight-cryptography/
+ * documents/finalist-round/updated-spec-doc/ascon-spec-final.pdf
+ */
+#define ASCONPS(x0, x1, x2, x3, x4) \
+    do {                            \
+        uint64_t q0, q1;            \
+                                    \
+        x0 ^= x4;                   \
+        x4 ^= x3;                   \
+        x2 ^= x1;                   \
+        q0 = x0 & (~x4);            \
+        q1 = x2 & (~x1);            \
+        x0 ^= q1;                   \
+        q1 = x4 & (~x3);            \
+        x2 ^= q1;                   \
+        q1 = x1 & (~x0);            \
+        x4 ^= q1;                   \
+        q1 = x3 & (~x2);            \
+        x1 ^= q1;                   \
+        x3 ^= q0;                   \
+        x1 ^= x0;                   \
+        x3 ^= x2;                   \
+        x0 ^= x4;                   \
+        x2 = ~x2;                   \
+    } while (0)
+
+/* linear layer, NIST SP 800-232 Figure 3 */
+#define ASCONPL(x0, x1, x2, x3, x4)          \
+    do {                                     \
+        x0 ^= ROR64(x0, 19) ^ ROR64(x0, 28); \
+        x1 ^= ROR64(x1, 61) ^ ROR64(x1, 39); \
+        x2 ^= ROR64(x2, 1) ^ ROR64(x2, 6);   \
+        x3 ^= ROR64(x3, 10) ^ ROR64(x3, 17); \
+        x4 ^= ROR64(x4, 7) ^ ROR64(x4, 41);  \
+    } while (0)
+
+/* one round */
+#define ASCONP1(x0, x1, x2, x3, x4, rcon)  \
+    do {                                   \
+        ASCONPC(x0, x1, x2, x3, x4, rcon); \
+        ASCONPS(x0, x1, x2, x3, x4);       \
+        ASCONPL(x0, x1, x2, x3, x4);       \
+    } while (0)
+
+/* 8 rounds */
+#define ASCONP8(x0, x1, x2, x3, x4)           \
+    do {                                      \
+        ASCONP1(x0, x1, x2, x3, x4, 0xB4ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0xA5ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x96ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x87ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x78ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x69ULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x5AULL); \
+        ASCONP1(x0, x1, x2, x3, x4, 0x4BULL); \
+    } while (0)
+
+/* 12 rounds */
+#define ASCONP12(x0, x1, x2, x3, x4)          \
+    do {                                      \
+        ASCONP1(x0, x1, x2, x3, x4, 0xF0ULL); \
```

Comment:
I must be not understanding the NIST specification
````
s2 ^= C(i)
````
Where i = round, and C(0) = 0x3B and C(4) = 0x4B...
Why are you starting at C(4) ???
Is this some difference between the original spec and NIST or am I misinterpreting?

This is why I am a bit paranoid about your test vectors...